### ISSUE TITLE
Variablen Deklaration überarbeiten 

### ISSUE THREAD

**Issue by jenuk** [FachschaftMathPhysInfo/Programmiervorkurs#78]
Man sollte Variablen nur einzeln deklarieren und erst, wenn man sie dann auch wirklich zuweist.

**Comment by friedz** [comment]
ist natürlich nicht ganz einfach wenn so viele über `std::cin >> ` eingelesen werden aber sollten wir möglichst anpassen

**Comment by jenuk** [comment]
Ist
```
int a = 0;
std::cin >> a;
```
schöner als
```
int a;
std::cin >> a;
```
Dann könnte man das gegebenfalls so ersetzen.

**Comment by fschledorn** [comment]
> Nach dem Stand der IPK und IPI atm wird glaube ich 
> ```
> int a;
> std::cin >> a;
> ```
> gelehrt. Ich wäre im Generellen dafür, dass wir uns an deren Vorschriften halten, da wir die Studis ja auf diese VLs vorbereiten wollen

Ja bzw, so kreierst du ggf. undefiniertes Verhalten wenn `a` nicht global oder static ist und sie ausgelesen wird bevor ein Wert zugewiesen wird. Besser (und moderner) wäre probs:

```
int a{};
```

**Comment by Plebysnacc** [comment]
Hmmm, lass uns die Woche mal schauen wie die letzte VL das gemacht hat, glaube das vorzugreifen ohne es zu erklären ist kontraproduktiv und es zu erklären überfordert eventuell das allg. geringe Niveau das wir ja gezielt anstreben

**Comment by fschledorn** [comment]
Yup, hätte ihnen ggf. einfach nur die Methode mit {} vorgestellt, dann lernen sie keine schlechten Gewohnheiten

**Comment by christian-heusel** [comment]
Aber dann ist es halt auch mehr als eine deklaration, da `{}` bei trivialen types AFAIU as zero-initializer funktioniert 😊 

> If T is a [scalar type](https://en.cppreference.com/w/cpp/named_req/ScalarType), the object is initialized to the value obtained by [explicitly converting](https://en.cppreference.com/w/cpp/language/explicit_cast) the integer literal ​0​ (zero) to T.

https://en.cppreference.com/w/cpp/language/zero_initialization

**Comment by fschledorn** [comment]
> Aber dann ist es halt auch mehr als eine deklaration, da `{}` bei trivialen types AFAIU as zero-initializer funktioniert 😊
> 
> > If T is a [scalar type](https://en.cppreference.com/w/cpp/named_req/ScalarType), the object is initialized to the value obtained by [explicitly converting](https://en.cppreference.com/w/cpp/language/explicit_cast) the integer literal ​0​ (zero) to T.
> 
> https://en.cppreference.com/w/cpp/language/zero_initialization

I know, das is ja genau mein Punkt. Im letzten IPK war die Idee am Anfang erst die safen Methoden vorzustellen und dann wenn sie sich daran gewöhnt haben erst die “unsafe” Sachen bei zu bringen. Also dass der Default Wert im Kopf der Erstis safe ist und erst danach zeigen wir Ihnen wann und warum sie davon abweichen können. 
Also mehr ne “pädagogische” Idee wie man dafür sorgt dass sie safe code schreiben(wenn auch nur accidentally).

**Comment by fschledorn** [comment]
Ach das meinst du. Hatten wir im IPK dieses Jahr aber auch ähnlich gehandhabt wenn ich mich richtig erinnere. Wir haben kurz eingeführt Deklaration vs Initialisierung. Und danach direkt gesagt: macht es, falls ihr keinen spezifischen Grund habt es zu trennen, in einem und dann über die `{}` Syntax. 
In den meisten Fällen war für uns das produzieren von undefiniertem Verhalten ein schlimmerer/größerer Fehler als nen falscher Wert.
Wollen wir das hier so ähnlich machen?

Der approach dieses Jahr war ziemlich stark darauf ausgelegt, dass wir “safes” C++ beibringen und nicht ultra Performantes. Haben z.B. auch gar keine Raw ptr gemacht sondern ausschließlich smart pointer mit einer ähnlichen Begründung.

**Comment by christian-heusel** [comment]
Ja ich finde die variante mit `{}` aus den oben genannten Gründen auch besser 😊
